### PR TITLE
[spruce] Bump @pinecone-database version to support global control plane

### PR DESCRIPTION
**Pull request by austin-denoble**
## Problem
We've been working to support the new global control plane service in the Pinecone clients. We need to update our sample apps to migrate to the new client versions, and update associated code paths as needed.

## Solution
- Update `README.md`, `.env.example`, and `validateEnvironmentVariables()` to remove `PINECONE_ENVIRONMENT`.
- Update `index.ts` to handle the new `listIndexes()` response shape, and passing the proper config to `createIndex()`. Update relevant snippets in `README.md`.

I also took a pass at some minor cleanup not directly related to the upgrade:
- Ran prettier across the codebase, some stuff was oddly formatted.
- Using `validateEnvironmentVariables()` relevant code is run (`index.ts`, and `recommend.ts`). This is minor as `getEnv()` will throw in the same way, but the validation util wasn't used at all.
- Removed duplicate `sliceIntoChunks()` function.
- Fixed an apparent issue in `chunkedUpsert()` where `chunks.map(async (chunk) => ...)` was upserting `vectors` and not the specific `chunk`.

For now, I've created a spruce branch to merge this into as we probably don't want to release sample apps that aren't aligned with what's available publicly.

## Type of Change
- [X] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [X] This change requires a documentation update

## Test Plan
There are no unit tests associated with this project, so testing manually is required. I tested both the embedding and upsert path along with the recommendations path, and they seem to work as we'd expect. We should add some level of testing and validation to this app at some point.

### Validate embedding and upsert
`npm run index`
Screenshot 2023-12-20 at 2 13 24 PM

### Validate recommendations
`npm run recommend -- --query="Xbox" --section="Games
Screenshot 2023-12-20 at 2 14 04 PM
"`

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -1,26 +1,34 @@
...
 import * as dfd from "danfojs-node";
 import { embedder } from "embeddings.ts";
 import loadCSVFile from "utils/csvLoader.ts";
 import splitFile from "utils/fileSplitter.ts";
-import { chunkedUpsert } from './utils/chunkedUpsert.ts';
 import type { ArticleRecord } from "types.ts";
+import { chunkedUpsert } from "./utils/chunkedUpsert.ts";
 
 dotenv.config();
+validateEnvironmentVariables();
```
Comment (austin-denoble) [inline src/index.ts:20]: Added validation here.

File: src/index.ts [hunk 2]
```diff
@@ -75,16 +91,22 @@ try {
 
   // Create the index if it doesn't already exist
   const indexList = await pinecone.listIndexes();
-  if (indexList.indexOf({ name: indexName }) === -1) {
-    await pinecone.createIndex({ name: indexName, dimension: 384, waitUntilReady: true })
+  if (!indexList.indexes?.some((index) => index.name === indexName)) {
```
Comment (austin-denoble) [inline src/index.ts:100]: Update to handle `listIndexes()` response and the `createIndex()` call below now includes `spec`.

File: src/recommend.ts [hunk 3]
```diff
@@ -1,53 +1,62 @@
...
+
 const indexName = getEnv("PINECONE_INDEX");
 const pinecone = new Pinecone();
 
 // Ensure the index exists
 try {
   const description = await pinecone.describeIndex(indexName);
   if (!description.status?.ready) {
-    throw `Index not ready, description was ${JSON.stringify(description)}`
+    throw new Error(
```
Comment (austin-denoble) [inline src/recommend.ts:22]: Lint was complaining about not throwing an `Error` here.

File: src/utils/chunkedUpsert.ts [hunk 4]
```diff
@@ -1,34 +1,29 @@
-import type { Index, PineconeRecord } from '@pinecone-database/pinecone';
+import type { Index, PineconeRecord } from "@pinecone-database/pinecone";
+import { sliceIntoChunks } from "./util.ts";
```
Comment (austin-denoble) [inline src/utils/chunkedUpsert.ts:2]: The `sliceIntoChunks` function below was a duplicate of what we have in utils.

File: src/utils/chunkedUpsert.ts [hunk 5]
```diff
@@ -1,34 +1,29 @@
...
+) => {
+  // Split the vectors into chunks
+  const chunks = sliceIntoChunks<PineconeRecord>(vectors, chunkSize);
+
+  try {
+    // Upsert each chunk of vectors into the index
+    await Promise.allSettled(
+      chunks.map(async (chunk) => {
+        try {
+          await index.namespace(namespace).upsert(chunk);
```
Comment (austin-denoble) [inline src/utils/chunkedUpsert.ts:18]: Previously we were upserting `vectors` instead of the chunk itself.

**Review by rschwabco — CHANGES_REQUESTED** [review]
Other than the hard coded region/cloud - lgtm

**Comment by austin-denoble** [comment]
> Other than the hard coded region/cloud - lgtm

`PINECONE_CLOUD` and `PINECONE_REGION` have been added as env variables.

**Pull request merged**